dandiapi/api/models/audit.py

```diff
@@ -0,0 +1,198 @@
+    def update_asset(*, dandiset: Dandiset, user: User, asset: Asset) -> AuditRecord:
+        details = AuditRecord._asset_details(asset)
+        return AuditRecord.make_audit_record(
+            dandiset=dandiset, user=user, record_type='update_asset', details=details
+        )
+
+    @staticmethod
+    def remove_asset(*, dandiset: Dandiset, user: User, asset: Asset) -> AuditRecord:
+        details = {
+            'path': asset.path,
+            'asset_id': asset.id,
+        }
+        return AuditRecord.make_audit_record(
+            dandiset=dandiset, user=user, record_type='remove_asset', details=details
+        )
+
+    @staticmethod
+    def create_zarr(*, dandiset: Dandiset, user: User, zarr_archive: ZarrArchive) -> AuditRecord:
+        details = {
+            'zarr_id': zarr_archive.id,
+            'name': zarr_archive.name,
+        }
+        return AuditRecord.make_audit_record(
+            dandiset=dandiset, user=user, record_type='create_zarr', details=details
+        )
+
+    @staticmethod
+    def upload_zarr(
+        *, dandiset: Dandiset, user: User, zarr_archive: ZarrArchive, urls: list[str]
+    ) -> AuditRecord:
+        details = {
+            'zarr_id': zarr_archive.id,
+            'num_urls': len(urls),
+        }
+        return AuditRecord.make_audit_record(
+            dandiset=dandiset, user=user, record_type='upload_zarr', details=details
+        )
+
+    @staticmethod
+    def delete_zarr_chunks(
+        *, dandiset: Dandiset, user: User, zarr_archive: ZarrArchive, paths: list[str]
+    ) -> AuditRecord:
+        details = {
+            'zarr_id': zarr_archive.id,
+            'num_chunks': len(paths),
+        }
+        return AuditRecord.make_audit_record(
+            dandiset=dandiset, user=user, record_type='delete_zarr_chunks', details=details
+        )
+
+    @staticmethod
+    def finalize_zarr(*, dandiset: Dandiset, user: User, zarr_archive: ZarrArchive) -> AuditRecord:
+        details = {
+            'zarr_id': zarr_archive.id,
+        }
+        return AuditRecord.make_audit_record(
+            dandiset=dandiset, user=user, record_type='finalize_zarr', details=details
+        )
+
+    @staticmethod
+    def unembargo_dandiset(*, dandiset: Dandiset, user: User) -> AuditRecord:
+        return AuditRecord.make_audit_record(
+            dandiset=dandiset, user=user, record_type='unembargo_dandiset', details={}
+        )
+
+    @staticmethod
+    def publish_dandiset(*, dandiset: Dandiset, user: User, version: str) -> AuditRecord:
+        details = {
+            'version': version,
+        }
+        return AuditRecord.make_audit_record(
+            dandiset=dandiset, user=user, record_type='publish_dandiset', details=details
+        )
+
+    @staticmethod
+    def delete_dandiset(*, dandiset: Dandiset, user: User):
+        return AuditRecord.make_audit_record(
+            dandiset=dandiset, user=user, record_type='delete_dandiset', details={}
+        )
```

Comment:
One way I thought about this is comparing it to publishing. In that case, we have a `publish` service that contains all the methods related to publishing. If we wanted to, we could have instead made a `publish` static method on the `Dandiset` class. I think `staticmethods` are a perfectly valid way to go about this, but it feels like making this a service would be more consistent with the rest of the codebase. Anyway, I'm interested to hear what Roni thinks too.

Reply:
I agree that converting my `staticmethod`s to service functions is the right thing to do, since that's the prevailing pattern in our codebase.

It's *obiter dicta* but fwiw, I agree that static methods would be a perfectly valid way of doing this type of thing, but also that service functions are equally valid. I like the idea that those methods are carried on the same class that they address, but that is also the idea behind a service module.

Reply:
Things got a bit sloppy, but the following commits implement a service layer for audit:
- e120f861a9f2cfad5024a1695eb9a8cfe4dece7b
- 6e4c2bb964aa6b3ac55c8a567591cef06e40a23b
- 2c2ce5d74b82ab971aadf0a878e1cd49b8eb5852